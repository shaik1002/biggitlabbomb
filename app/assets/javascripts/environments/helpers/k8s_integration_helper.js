import csrf from '~/lib/utils/csrf';
import { CLUSTER_AGENT_ERROR_MESSAGES } from '../constants';

export function humanizeClusterErrors(reason) {
  const errorReason = String(reason).toLowerCase();
  const errorMessage = CLUSTER_AGENT_ERROR_MESSAGES[errorReason];
  return errorMessage || CLUSTER_AGENT_ERROR_MESSAGES.other;
}

export function createK8sAccessConfiguration({ kasTunnelUrl, gitlabAgentId }) {
  return {
    basePath: kasTunnelUrl,
    headers: {
      'GitLab-Agent-Id': gitlabAgentId,
      'Content-Type': 'application/json',
      Accept: 'application/json',
      ...csrf.headers,
    },
    credentials: 'include',
  };
}
