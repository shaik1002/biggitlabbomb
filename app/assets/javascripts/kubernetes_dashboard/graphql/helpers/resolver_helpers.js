import {
  CoreV1Api,
  Configuration,
  WatchApi,
  EVENT_DATA,
  EVENT_TIMEOUT,
  EVENT_ERROR,
} from '@gitlab/cluster-client';
import { connectionStatus } from '~/environments/graphql/resolvers/kubernetes/constants';
import { updateConnectionStatus } from '~/environments/graphql/resolvers/kubernetes/k8s_connection_status';
import { s__ } from '~/locale';

export const handleClusterError = async (err) => {
  if (!err.response) {
    throw err;
  }

  const contentType = err.response.headers.get('Content-Type');

  if (contentType !== 'application/json') {
    throw new Error(
      s__(
        'KubernetesDashboard|There was a problem fetching cluster information. Refresh the page and try again.',
      ),
    );
  }

  const errorData = await err.response.json();
  throw errorData;
};

export const buildWatchPath = ({ resource, api = 'api/v1', namespace = '' }) => {
  return namespace ? `/${api}/namespaces/${namespace}/${resource}` : `/${api}/${resource}`;
};

export const mapWorkloadItem = (item) => {
  const status = item.status || {};
  const spec = item.spec || {};
  const metadata = {
    ...item.metadata,
    annotations: item.metadata?.annotations || {},
    labels: item.metadata?.labels || {},
  };
  return { status, spec, metadata, __typename: 'LocalWorkloadItem' };
};

export const watchWorkloadItems = ({
  client,
  query,
  configuration,
  namespace,
  watchPath,
  queryField,
}) => {
  const config = new Configuration(configuration);
  const watcherApi = new WatchApi(config);

  updateConnectionStatus(client, {
    configuration,
    namespace,
    resourceType: queryField,
    status: connectionStatus.connecting,
  });

  watcherApi
    .subscribeToStream(watchPath, { watch: true })
    .then((watcher) => {
      let result = [];

      watcher.on(EVENT_DATA, (data) => {
        result = data.map(mapWorkloadItem);
        client.writeQuery({
          query,
          variables: { configuration, namespace },
          data: { [queryField]: result },
        });
        updateConnectionStatus(client, {
          configuration,
          namespace,
          resourceType: queryField,
          status: connectionStatus.connected,
        });
      });

      watcher.on(EVENT_TIMEOUT, () => {
        result = [];

        updateConnectionStatus(client, {
          configuration,
          namespace,
          resourceType: queryField,
          status: connectionStatus.disconnected,
        });
      });

      watcher.on(EVENT_ERROR, () => {
        result = [];
        updateConnectionStatus(client, {
          configuration,
          namespace,
          resourceType: queryField,
          status: connectionStatus.disconnected,
        });
      });
    })
    .catch((err) => {
      handleClusterError(err);
    });
};

export const getK8sPods = ({
  client,
  query,
  configuration,
  namespace = '',
  mapFn = mapWorkloadItem,
  queryField = 'k8sPods',
}) => {
  const config = new Configuration(configuration);

  const coreV1Api = new CoreV1Api(config);
  const podsApi = namespace
    ? coreV1Api.listCoreV1NamespacedPod({ namespace })
    : coreV1Api.listCoreV1PodForAllNamespaces();

  return podsApi
    .then((res) => {
      const watchPath = buildWatchPath({ resource: 'pods', namespace });
      watchWorkloadItems({
        client,
        query,
        configuration,
        namespace,
        watchPath,
        queryField,
      });

      const data = res?.items || [];
      return data.map(mapFn);
    })
    .catch(async (err) => {
      try {
        await handleClusterError(err);
      } catch (error) {
        throw new Error(error.message);
      }
    });
};
