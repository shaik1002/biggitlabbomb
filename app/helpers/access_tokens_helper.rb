# frozen_string_literal: true

module AccessTokensHelper
  include AccountsHelper
  include ApplicationHelper

  def scope_description(prefix)
    case prefix
    when :project_access_token
      [:doorkeeper, :project_access_token_scope_desc]
    when :group_access_token
      [:doorkeeper, :group_access_token_scope_desc]
    else
      [:doorkeeper, :scope_desc]
    end
  end

  def tokens_app_data
    {
      feed_token: {
        enabled: !Gitlab::CurrentSettings.disable_feed_token,
        token: current_user.feed_token,
        reset_path: reset_feed_token_profile_path
      },
      incoming_email_token: {
        enabled: incoming_email_token_enabled?,
        token: current_user.enabled_incoming_email_token,
        reset_path: reset_incoming_email_token_profile_path
      },
      static_object_token: {
        enabled: static_objects_external_storage_enabled?,
        token: current_user.enabled_static_object_token,
        reset_path: reset_static_object_token_profile_path
      }
    }.to_json
  end

  def expires_at_field_data
    {
      min_date: 1.day.from_now.iso8601,
      max_date: PersonalAccessToken::MAX_PERSONAL_ACCESS_TOKEN_LIFETIME_IN_DAYS.days.from_now.iso8601
    }
  end

  def show_token_expiration_banner?
    Gitlab::Auth::TokenExpirationBanner.show_token_expiration_banner?
  end
end

AccessTokensHelper.prepend_mod
