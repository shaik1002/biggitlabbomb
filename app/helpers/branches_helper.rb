# frozen_string_literal: true

module BranchesHelper
  def project_branches
    options_for_select(@project.repository.branch_names, @project.default_branch)
  end

  def protected_branch?(project, branch)
    ProtectedBranch.protected?(project, branch.name)
  end

  def access_levels_data(access_levels)
    return [] unless access_levels

    access_levels.map do |level|
      if level.type == :deploy_key
        { id: level.id, type: level.type, deploy_key_id: level.deploy_key_id }
      else
        { id: level.id, type: :role, access_level: level.access_level }
      end
    end
  end

  def merge_request_status(merge_request)
    return unless merge_request.present?
    return if merge_request.closed?

    if merge_request.open? || merge_request.locked?
      variant = :success
      variant = :warning if merge_request.draft?

      mr_icon = 'merge-request-open'
      mr_status = _('Open')
    elsif merge_request.merged?
      variant = :info
      mr_icon = 'merge'
      mr_status = _('Merged')
    else
      return
    end

    { icon: mr_icon, title: "#{mr_status} - #{merge_request.title}", variant: variant }
  end
end

BranchesHelper.prepend_mod_with('BranchesHelper')
