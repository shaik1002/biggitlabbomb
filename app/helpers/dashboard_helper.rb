# frozen_string_literal: true

module DashboardHelper
  include IconsHelper

  def has_start_trial?
    false
  end

  def feature_entry(title, href: nil, enabled: true, doc_href: nil)
    enabled_text = enabled ? 'on' : 'off'
    label = "#{title}: status #{enabled_text}"
    link_or_title = href && enabled ? tag.a(title, href: href) : title

    tag.p(aria: { label: label }) do
      concat(link_or_title)

      concat(tag.span(class: %w[light gl-float-right]) do
        boolean_to_icon(enabled)
      end)

      if doc_href.present?
        link_to_doc = link_to(
          sprite_icon('question-o'),
          doc_href,
          class: 'gl-ml-2',
          title: _('Documentation'),
          target: '_blank',
          rel: 'noopener noreferrer'
        )

        concat(link_to_doc)
      end
    end
  end

  def user_groups_requiring_reauth
    []
  end
end

DashboardHelper.prepend_mod_with('DashboardHelper')
