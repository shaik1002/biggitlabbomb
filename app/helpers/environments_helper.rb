# frozen_string_literal: true

module EnvironmentsHelper
  include ActionView::Helpers::AssetUrlHelper

  def environments_list_data
    {
      endpoint: project_environments_path(@project, format: :json)
    }
  end

  def environments_folder_list_view_data(project, folder)
    {
      "endpoint" => folder_project_environments_path(project, folder, format: :json),
      "folder_name" => folder,
      "project_path" => project.full_path,
      "help_page_path" => help_page_path("ci/environments/index"),
      "can_read_environment" => can?(current_user, :read_environment, @project).to_s
    }
  end

  def custom_metrics_available?(project)
    can?(current_user, :admin_project, project)
  end

  def metrics_data(project, environment)
    return {} if Feature.enabled?(:remove_monitor_metrics)

    metrics_data = {}
    metrics_data.merge!(project_metrics_data(project)) if project
    metrics_data.merge!(environment_metrics_data(environment)) if environment
    metrics_data.merge!(project_and_environment_metrics_data(project, environment)) if project && environment
    metrics_data.merge!(static_metrics_data)

    metrics_data
  end

  def can_destroy_environment?(environment)
    can?(current_user, :destroy_environment, environment)
  end

  private

  def project_metrics_data(project)
    return {} unless project

    {
      'settings_path' => edit_project_settings_integration_path(project, 'prometheus'),
      'clusters_path' => project_clusters_path(project),
      'default_branch' => project.default_branch,
      'project_path' => project_path(project),
      'tags_path' => project_tags_path(project),
      'custom_metrics_path' => project_prometheus_metrics_path(project),
      'validate_query_path' => validate_query_project_prometheus_metrics_path(project),
      'custom_metrics_available' => custom_metrics_available?(project).to_s
    }
  end

  def environment_metrics_data(environment)
    return {} unless environment

    {
      'current_environment_name' => environment.name,
      'has_metrics' => environment.has_metrics?.to_s,
      'environment_state' => environment.state.to_s
    }
  end

  def project_and_environment_metrics_data(project, environment)
    return {} unless project && environment

    {
      'deployments_endpoint' => project_environment_deployments_path(project, environment, format: :json),
      'operations_settings_path' => project_settings_operations_path(project),
      'can_access_operations_settings' => can?(current_user, :admin_operations, project).to_s
    }
  end

  def static_metrics_data
    {
      'documentation_path' => help_page_path('administration/monitoring/prometheus/index'),
      'add_dashboard_documentation_path' => help_page_path('operations/metrics/dashboards/index', anchor: 'add-a-new-dashboard-to-your-project')
    }
  end
end

EnvironmentsHelper.prepend_mod_with('EnvironmentsHelper')
