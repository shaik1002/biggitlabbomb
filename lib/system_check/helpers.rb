# frozen_string_literal: true

module SystemCheck
  # Helpers used inside a SystemCheck instance to standardize output responses
  module Helpers
    include ::Gitlab::TaskHelpers

    # Display a message telling to fix and rerun the checks
    def fix_and_rerun
      $stdout.puts '  Please fix the error above and rerun the checks.'.color(:red)
    end

    # Display a formatted list of references (documentation or links) where to find more information
    #
    # @param [Array<String>] sources one or more references (documentation or links)
    def for_more_information(*sources)
      $stdout.puts '  For more information see:'.color(:blue)
      sources.each do |source|
        $stdout.puts "  #{source}"
      end
    end

    # Construct a help page based on the instance's external_url
    #
    # @param <String> path of the documentation page
    # @param <String> anchor to the specific docs heading
    #
    # @return <String> URL of the help page
    def construct_help_page_url(path, anchor = nil)
      Rails.application.routes.url_helpers.help_page_url(path, anchor)
    end

    def see_installation_guide_section(section)
      "doc/install/installation.md in section \"#{section}\""
    end

    # @deprecated This will no longer be used when all checks were executed using SystemCheck
    def finished_checking(component)
      $stdout.puts ''
      $stdout.puts "Checking #{component.color(:yellow)} ... #{'Finished'.color(:green)}"
      $stdout.puts ''
    end

    # @deprecated This will no longer be used when all checks were executed using SystemCheck
    def start_checking(component)
      $stdout.puts "Checking #{component.color(:yellow)} ..."
      $stdout.puts ''
    end

    # Display a formatted list of instructions on how to fix the issue identified by the #check?
    #
    # @param [Array<String>] steps one or short sentences with help how to fix the issue
    def try_fixing_it(*steps)
      steps = steps.shift if steps.first.is_a?(Array)

      $stdout.puts '  Try fixing it:'.color(:blue)
      steps.each do |step|
        $stdout.puts "  #{step}"
      end
    end

    def sanitized_message(project)
      if should_sanitize?
        "#{project.namespace_id.to_s.color(:yellow)}/#{project.id.to_s.color(:yellow)} ... "
      else
        "#{project.full_name.color(:yellow)} ... "
      end
    end

    def should_sanitize?
      ENV['SANITIZE'] == 'true'
    end

    def omnibus_gitlab?
      Dir.pwd == '/opt/gitlab/embedded/service/gitlab-rails'
    end

    def sudo_gitlab(command)
      "sudo -u #{gitlab_user} -H #{command}"
    end
  end
end
