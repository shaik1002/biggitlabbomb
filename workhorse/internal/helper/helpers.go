package helper

import (
	"errors"
	"mime"
	"net/url"
	"os"

	"gitlab.com/gitlab-org/gitlab/workhorse/internal/log"
)

func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(path)
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.WithError(err).WithFields(log.Fields{"url": s}).Fatal("urlMustParse")
	}
	return u
}

func IsContentType(expected, actual string) bool {
	parsed, _, err := mime.ParseMediaType(actual)
	return err == nil && parsed == expected
}
