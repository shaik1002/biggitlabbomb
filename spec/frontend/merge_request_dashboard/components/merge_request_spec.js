import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import MergeRequest from '~/merge_request_dashboard/components/merge_request.vue';
import CiIcon from '~/vue_shared/components/ci_icon/ci_icon.vue';

describe('Merge request dashboard merge request component', () => {
  let wrapper;

  function createComponent(mergeRequest = {}) {
    wrapper = shallowMountExtended(MergeRequest, {
      propsData: {
        mergeRequest: {
          reference: '!123456',
          titleHtml: 'Merge request title',
          author: {
            name: 'John Smith',
            webUrl: 'https://gitlab.com/root',
          },
          milestone: {
            title: '17.0',
          },
          labels: {
            nodes: [
              {
                id: 'gid://gitlab/GroupLabel/992791',
                color: '#428BCA',
                title: 'Deliverable',
                description: 'Label description',
              },
            ],
          },
          assignees: {
            nodes: [
              {
                id: 'gid://gitlab/User/1',
                avatarUrl: '',
                name: 'John Smith',
                username: 'jsmith',
                webUrl: 'https://gitlab.com/root',
                webPath: '/root',
              },
            ],
          },
          reviewers: {
            nodes: [
              {
                id: 'gid://gitlab/User/1',
                avatarUrl: '',
                name: 'John Smith',
                username: 'jsmith',
                webUrl: 'https://gitlab.com/root',
                webPath: '/root',
              },
              {
                id: 'gid://gitlab/User/2',
                avatarUrl: '',
                name: 'John Smith',
                username: 'jsmith',
                webUrl: 'https://gitlab.com/root',
                webPath: '/root',
              },
            ],
          },
          userDiscussionsCount: 5,
          createdAt: '2024-04-22T10:13:09Z',
          updatedAt: '2024-04-19T14:34:42Z',
          ...mergeRequest,
        },
      },
    });
  }

  it('renders template', () => {
    createComponent();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders CI icon for headPipeline', () => {
    createComponent({
      headPipeline: {
        id: 'gid://gitlab/Ci::Pipeline/1',
        detailedStatus: {
          id: 'success-1',
          icon: 'status_success',
          text: 'Passed',
          detailsPath: '/',
        },
      },
    });

    expect(wrapper.findComponent(CiIcon).exists()).toBe(true);
    expect(wrapper.findComponent(CiIcon).props('status')).toEqual({
      id: 'success-1',
      icon: 'status_success',
      text: 'Passed',
      detailsPath: '/',
    });
  });
});
