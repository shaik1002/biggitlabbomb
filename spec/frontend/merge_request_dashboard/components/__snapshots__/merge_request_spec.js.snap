// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge request dashboard merge request component renders template 1`] = `
<div
  class="gl-bg-white gl-p-5 gl-rounded-base"
>
  <div
    class="gl-display-flex"
  >
    <div
      class="gl-display-flex gl-flex-direction-column"
    >
      <h4
        class="gl-font-base gl-mb-0 gl-mt-0"
      >
        <gl-link-stub
          class="gl-hover-text-gray-900 gl-text-body"
        >
          Merge request title
        </gl-link-stub>
      </h4>
      <div
        class="gl-font-sm gl-mt-2 gl-text-secondary"
      >
        <gl-sprintf-stub
          message="%{reference} %{divider} created %{createdAt} by %{author} %{milestone}"
        />
      </div>
    </div>
    <div
      class="gl-display-flex gl-flex-direction-column gl-ml-auto"
    >
      <ul
        class="gl-display-flex gl-justify-content-end gl-list-none gl-m-0 gl-p-0"
      >
        <li
          class="gl-ml-4"
        >
          <user-avatar-link-stub
            imgalt=""
            imgcssclasses=""
            imgcsswrapperclasses=""
            imgsize="24"
            imgsrc=""
            linkhref="https://gitlab.com/root"
            popoveruserid=""
            popoverusername=""
            tooltipplacement="top"
            tooltiptext="Assigned to John Smith"
            username=""
          />
        </li>
        <li
          class="gl-ml-4"
        >
          <user-avatar-link-stub
            class="gl-mr-2"
            imgalt=""
            imgcssclasses=""
            imgcsswrapperclasses=""
            imgsize="24"
            imgsrc=""
            linkhref="https://gitlab.com/root"
            popoveruserid=""
            popoverusername=""
            tooltipplacement="top"
            tooltiptext="Review requested John Smith"
            username=""
          />
          <user-avatar-link-stub
            imgalt=""
            imgcssclasses=""
            imgcsswrapperclasses=""
            imgsize="24"
            imgsrc=""
            linkhref="https://gitlab.com/root"
            popoveruserid=""
            popoverusername=""
            tooltipplacement="top"
            tooltiptext="Review requested John Smith"
            username=""
          />
        </li>
        <li
          class="gl-align-self-center gl-ml-4"
        >
          <gl-icon-stub
            class="!gl-align-middle"
            name="comments"
            size="16"
          />
          5
        </li>
      </ul>
      <div
        class="gl-font-sm gl-mt-2 gl-text-right gl-text-secondary"
      >
        <gl-sprintf-stub
          message="Updated at %{updatedAt}"
        />
      </div>
    </div>
  </div>
  <div>
    <gl-label-stub
      backgroundcolor="#428BCA"
      class="gl-mr-2"
      description="Label description"
      size="sm"
      target=""
      title="Deliverable"
      tooltipplacement="top"
    />
  </div>
</div>
`;
